src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -45,15 +44,20 @@ public Factory()
         private const string RazorSourceGeneratorTypeName = "Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator";
         private static readonly string s_razorSourceGeneratorFileNamePrefix = Path.Combine(RazorSourceGeneratorAssemblyName, RazorSourceGeneratorTypeName);
 
+        private readonly Workspace _workspace;
+
         private readonly object _gate = new();
 
-#if NETCOREAPP
-        private readonly ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSoution = new();
-#else
-        // Framework lacks both a .Clear() method.  So for Framework we simulate that by just overwriting this with a
-        // new instance.  This happens under a lock, so everyone sees a consistent dictionary.
-        private ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSoution = new();
-#endif
+        /// <summary>
+        /// Cached compile time solution corresponding to the <see cref="Workspace.PrimaryBranchId"/>
+        /// </summary>
+        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _primaryBranchCompileTimeCache;
+
+        /// <summary>
+        /// Cached compile time solution for a forked branch.  This is used primarily by LSP cases where
+        /// we fork the workspace solution and request diagnostics for the forked solution.
+        /// </summary>
+        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _forkedBranchCompileTimeCache;
```

Comment:
@tmat any concern with bringing this back? I don't know if anything more recently depended upon the simpler code in some way.

Reply:
I am not aware of anything. Why do we need to revert this?

Reply:
Because it caused a big regression of allocation in OOP
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1426911